nuxeo-platform-spreadsheet-web/src/main/js/app/ui/spreadsheet.js

```diff
@@ -57,29 +57,95 @@ class Spreadsheet {
     this.query.depth = 'max';
     // translate directory labels
     this.query.translate('directoryEntry', 'label');
+    this.query.pageProvider = pageProviderName;
 
-    this.query.pageProvider = pageProvider;
-
-    new Layout(connection, layout, language).fetch().then((layout) => {
-      // Check which columns to display
-      var cols = (!columns) ?
-          layout.columns.filter((c) =>  c.selectedByDefault !== false)
-          :
-          columns.map((name) => layout.columns.filter((c) => c.name === name)[0]);
-      this.columns = cols
+    // set columns based on result layout
+    if (resultLayoutName) {
+      new Layout(connection, resultLayoutName, language).fetch().then((layout) => {
+        // Check which columns to display
+        let cols = (!resultColumns) ? layout.columns.filter((c) => c.selectedByDefault !== false)
+          : resultColumns.map((name) => layout.columns.filter((c) => c.name === name)[0]);
+        this.columns = cols
           // Exclude columns without widgets
           .filter((c) => c.widgets)
           // Create our columns wrapper
           .map((c) => new Column(connection, c, layout.widgets[c.widgets[0].name], this.dirtyRenderer.bind(this)))
           // Only show columns with a known widget type and with a field
           .filter((c) => c.hasSupportedWidgetType && c.field);
-    });
+      });
+
+    // or based on result columns only
+    } else {
+
+      // default columns
+      if (!resultColumns) {
+        resultColumns = [
+          { label: 'Title', field: 'dc:title' },
+          { label: 'Modified', field: 'dc:modified'},
+          { label: 'Last Contributor', field: 'dc:lastContributor'},
+          { label: 'State', field: 'currentLifeCycleState'}
+        ];
+      }
+
+      // get schemas prefixes from columns
+      let schemasPrefixes = [];
+      for (let c of resultColumns) {
+        let schema = c.field.includes(':') ? c.field.split(':')[0] : undefined;
+        if (schema && schemasPrefixes.indexOf(schema) === -1) {
+          schemasPrefixes.push(schema);
+        }
+      }
+
+      // fetch schemas (based on prefixes)
+      new Schemas(connection).fetch(schemasPrefixes).then((schemas) => {
+
+        // set columns
+        let cols = [];
+        for (let col of resultColumns) {
+
+          let column = {};
+          column.def = {properties: {any: {sortPropertyName: col.field, label: col.label}}};
+          column.widget = {field: col.field};
+
+          // get field from schemas map
+          let field = undefined; // <- explicitly set field as undefined in each iteration
+          if (col.field.includes(':')) {
+            let s = col.field.split(':')[0];
+            let f = col.field.split(':')[1];
+            field = schemas[s].fields[f] || undefined;
+          }
+
+          // set column widget type and properties based on field constraints
+          if (field) {
```

Comment:
isn't field only defined when you enter the previous if block ? if so then I guess this should be in that same block... or just do continue  if (!col.field.includes(':'))